### PR TITLE
Keyvalue lookup support in actionchains

### PR DESCRIPTION
**Pull request by manasdk**
- 'utility' to do keyvalue lookup which support hierarchical key structure (kind of)
- support for key values stored in datastore in actionchains

### CODE REVIEW

File: st2reactor/st2reactor/rules/datatransform.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def _construct_system_context(data, context):
         """Identify the system context in the data."""
         # The following regex will look for all occurrences of "{{system.*}}",
         # "{{ system.* }}", "{{ system.*}}", and "{{system.* }}" in the data.
-        regex = '{{\s*' + SYSTEM_PREFIX + '.(.*?)\s*}}'
+        regex = '{{\s*' + SYSTEM_KV_PREFIX + '.(.*?)\s*}}'
```
Comment (Kami) [inline st2reactor/st2reactor/rules/datatransform.py]: The dot before `(` should be escaped, right?

File: st2common/st2common/services/keyvalues.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from st2common.persistence.datastore import KeyValuePair
+
+
+class KeyValueLookup(object):
+
+    def __init__(self, base_lookup='', cache=None):
```
Comment (Kami) [inline st2common/st2common/services/keyvalues.py]: I find `base_lookup` name kinda confusing. Looking at the code bellow, it acts as a prefix so `prefix` / `key_prefix` might be a better name?

Comment (lakshmi-kannan) [inline st2common/st2common/services/keyvalues.py]: Yeah, i didn't get it either.

File: st2common/st2common/services/keyvalues.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    def __getitem__(self, key):
+        return self._get(key)
+
+    def __getattr__(self, name):
+        return self._get(name)
+
+    def _get(self, name):
+        # get the value for this key and save in value_cache
+        lookup = name if len(self._base_lookup) == 0 else '%s.%s' % (self._base_lookup, name)
```
Comment (Kami) [inline st2common/st2common/services/keyvalues.py]:
```python
lookup =  '%s.%s' % (self._base_lookup, name) ir self._base_loopup else name
```
Might be a better more Pythonic and readable.

File: st2reactor/st2reactor/rules/datatransform.py [hunk 4]
```diff
@@ -69,9 +69,9 @@ def _construct_system_context(data, context):
             kvp = KeyValuePair.get_by_name(key)
             kvps[key] = kvp.value
         if kvps:
-            if SYSTEM_PREFIX not in context:
-                context[SYSTEM_PREFIX] = {}
-            context[SYSTEM_PREFIX].update(kvps)
+            if SYSTEM_KV_PREFIX not in context:
+                context[SYSTEM_KV_PREFIX] = {}
+            context[SYSTEM_KV_PREFIX].update(kvps)
```
Comment (lakshmi-kannan) [inline st2reactor/st2reactor/rules/datatransform.py]: this doesn't make sense.

File: st2reactor/st2reactor/rules/datatransform.py (hunk 1 above)
Comment (manasdk) [inline st2reactor/st2reactor/rules/datatransform.py]: I don't understand regex. If I have to fix this code I will replace with the keyvalue lookup :)

File: st2common/st2common/services/keyvalues.py (hunk 3 above)
Comment (manasdk) [inline st2common/st2common/services/keyvalues.py]: Fixed 2865885b8c706e6f6bbe6a70edec4d3bb637ca79

File: st2common/st2common/services/keyvalues.py (hunk 2 above)
Comment (manasdk) [inline st2common/st2common/services/keyvalues.py]: Fixed 2865885b8c706e6f6bbe6a70edec4d3bb637ca79

File: st2reactor/st2reactor/rules/datatransform.py (hunk 1 above)
Comment (manasdk) [inline st2reactor/st2reactor/rules/datatransform.py]: Fixed in c360723a3565d7dacd96434a53017471db2a94d1. No more regex; replaced with KeyValueLookup. This way we can support dotted properties.

File: st2reactor/st2reactor/rules/datatransform.py (hunk 4 above)
Comment (manasdk) [inline st2reactor/st2reactor/rules/datatransform.py]: Got rid of it entirely.

**Pull request merged**